wandb/internal/tb_watcher.py

```diff
@@ -0,0 +1,279 @@
+    util.mkdir_exists_ok(os.path.dirname(wandb_path))
+    # We overwrite existing symlinks because namespaces can change in Tensorboard
+    if os.path.islink(wandb_path) and abs_path != os.readlink(wandb_path):
+        os.remove(wandb_path)
+        os.symlink(abs_path, wandb_path)
+    elif not os.path.exists(wandb_path):
+        os.symlink(abs_path, wandb_path)
+    # TODO(jhr): need to figure out policy, live/throttled?
+    sender._save_file(file_name)
+
+
+class TBWatcher(object):
+    def __init__(self, settings, sender=None):
+        self._logdirs = {}
+        self._consumer = None
+        self._settings = settings
+        self._sender = sender
+        # TODO(jhr): do we need locking in this queue?
+        self._watcher_queue = queue.PriorityQueue()
+
+    def _calculate_namespace(self, logdir):
+        dirs = list(self._logdirs) + [logdir]
+        rootdir = os.path.dirname(os.path.commonprefix(dirs))
+        if os.path.isfile(logdir):
+            filename = os.path.basename(logdir)
+        else:
+            filename = ""
+        # Tensorboard loads all tfevents files in a directory and prepends
+        # their values with the path.  Passing namespace to log allows us
+        # to nest the values in wandb
+        namespace = logdir.replace(filename, "").replace(rootdir, "").strip(os.sep)
+        # TODO: revisit this heuristic, it exists because we don't know the
+        # root log directory until more than one tfevents file is written to
+        if len(dirs) == 1 and namespace not in ["train", "validation"]:
+            namespace = None
+        return namespace
+
+    def add(self, logdir, save):
+        if logdir in self._logdirs:
+            return
+        namespace = self._calculate_namespace(logdir)
+        # TODO(jhr): implement the deferred tbdirwatcher to find namespace
+
+        if not self._consumer:
+            self._consumer = TBEventConsumer(self, self._watcher_queue)
+            self._consumer.start()
+
+        tbdir_watcher = TBDirWatcher(self, logdir, save, namespace, self._watcher_queue)
+        self._logdirs[logdir] = tbdir_watcher
+        tbdir_watcher.start()
+
+    def finish(self):
+        for tbdirwatcher in six.itervalues(self._logdirs):
+            tbdirwatcher.shutdown()
+        for tbdirwatcher in six.itervalues(self._logdirs):
+            tbdirwatcher.finish()
+        if self._consumer:
+            self._consumer.finish()
+
+
+class TBDirWatcher(object):
+    from tensorboard.backend.event_processing import directory_watcher  # type: ignore
+    from tensorboard.backend.event_processing import event_file_loader  # type: ignore
+    from tensorboard.compat import tf as tf_compat  # type: ignore
+
+    def __init__(self, tbwatcher, logdir, save, namespace, queue):
+        self._tbwatcher = tbwatcher
+        self._generator = self.directory_watcher.DirectoryWatcher(
+            logdir, self._loader(save, namespace), self._is_new_tensorflow_events_file
```

Comment:
Something I haven't tried, but I think might *just work...  If `logdir` is a cloud storage url, the DirectoryWatcher uses the io.gfile implementation so I think it will poll cloud storage for changes.  We would need to make `_is_new_tensorflow_events_file` use [tf.io.exists(...)](https://github.com/tensorflow/tensorboard/blob/master/tensorboard/compat/tensorflow_stub/io/gfile.py).  We would also want to try to pull in real tensorflow in place of tf_compat because the stock tensorboard implementation only supports S3.

The boss move would be to make a `wandb.io` library that mirrored [tf.io](https://github.com/tensorflow/tensorflow/tree/v2.3.0/tensorflow/python/lib/io) and we start compiling wheels.  Would make artifact references much simpler and enable some cool usecases.

Reply:
Wanted to wait on that until I had more time test non local files.